src/linear_algebra/decomposition.lean

```diff
@@ -0,0 +1,263 @@
+rfl
+
+variables {R M}
+
+/--
+Given an decomposition `D`, build an equivalence between the direct sum of `D.factors` and `M`.
+-/
+def to_equiv (D : decomposition ι R M) : (⨁ i, D.factors i) ≃ₗ[R] M :=
+((direct_sum_equiv_of_independent D.factors D.factors_ind).trans
+  (linear_equiv.of_eq _ _ D.factors_supr)).trans (submodule.top_equiv_self)
+
+lemma to_equiv_apply (D : decomposition ι R M) (x : ⨁ i, D.factors i) :
+  D.to_equiv x = ∑ i in x.support, x i :=
+begin
+  rw [to_equiv],
+  rw linear_equiv.trans_apply,
+  rw linear_equiv.trans_apply,
+  simp [submodule.top_equiv_self],
+end
+
+lemma to_equiv_apply' (D : decomposition ι R M) (x : ⨁ i, D.factors i) :
+  D.to_equiv x = direct_sum.to_module R ι M (λ i, (D.factors i).subtype) x :=
+begin
+  rw to_equiv_apply,
+  rw submodule_is_internal.apply,
+end
+
+lemma to_equiv_symm_single_apply
+  (D : decomposition ι R M) (i : ι) (x : M) (hx : x ∈ D.factors i) :
+  D.to_equiv.symm x = single i ⟨x, hx⟩ :=
+begin
+  apply_fun D.to_equiv using linear_equiv.injective _,
+  rw [linear_equiv.apply_symm_apply],
+  rw to_equiv_apply,
+  by_cases h : x = 0,
+  { rw dfinsupp.support_single_eq_zero,
+    rwa finset.sum_empty,
+    rwa ← submodule.coe_eq_zero },
+  rw [support_single_ne_zero, finset.sum_singleton, single_eq_same, submodule.coe_mk],
+  refine λ h', h _,
+  rw [← submodule.coe_eq_zero.mpr h', submodule.coe_mk]
+end
+
+lemma is_internal (D : decomposition ι R M) :
+  direct_sum.submodule_is_internal D.factors :=
+begin
+  rw direct_sum.submodule_is_internal,
+  convert D.to_equiv.bijective using 1, ext,
+  rw to_equiv_apply',
+end
+
+/--
+Given an indexed collection of submodules of `M` and a proof that they form an internal
+direct sum, construct a decomposition.
+-/
+def of_is_internal (p : ι → submodule R M) (hp : direct_sum.submodule_is_internal p) :
+  decomposition ι R M :=
+{ factors := p,
+  factors_ind := begin
+    rw complete_lattice.independent_def,
+    refine λ i, submodule.disjoint_def.mpr (λ x hi hSup, _),
+    apply_fun (submodule_is_internal.to_equiv p hp).symm using linear_equiv.injective _,
+    rw linear_equiv.map_zero,
+    -- now we unfold the info from `hSup`
+    obtain ⟨v, hv, hsum, hzero⟩ := (mem_bsupr _ _).mp hSup,
+    simp only [forall_eq, not_not] at hzero,
+    apply_fun (submodule_is_internal.to_equiv p hp).symm at hsum,
+    rw linear_equiv.map_sum at hsum,
+    have key := λ i, submodule_is_internal.to_equiv_symm_single_apply p hp i (v i) (hv i),
+    simp only [key] at hsum,
+    -- do casework on `i = j`
+    ext j, by_cases h : i = j,
+    { rw [← h, ← hsum, direct_sum.zero_apply, coe_zero, coe_eq_zero, finset_sum_apply],
+      simp only [single_apply, finset.sum_dite_eq', finsupp.mem_support_iff, ne.def,
+        mk_eq_zero, not_imp_self, ite_eq_right_iff, hzero], },
+    rw [submodule_is_internal.to_equiv_symm_single_apply p hp i x hi, single_eq_of_ne], refl,
+    exact h
+  end,
+  factors_supr := begin
```

Comment:
I've PR'd a shorter proof of this in #8274, which also eliminates the need for `mem_supr`

Reply:
I went ahead and pushed a commit to this PR that uses the new lemma.